ipld/read.go

```diff
@@ -26,21 +26,57 @@ func RetrieveData(
 	codec rsmt2d.Codec,
 ) (*rsmt2d.ExtendedDataSquare, error) {
 	edsWidth := len(dah.RowsRoots)
-	rowRoots := dah.RowsRoots
 	dataSquare := make([][]byte, edsWidth*edsWidth)
 
-	quadrant, err := pickRandomQuadrant(rowRoots)
+	q := rand.Intn(4) + 1 //nolint:gosec
+	quadrant, err := pickQuadrant(q, dah.RowsRoots)
 	if err != nil {
 		return nil, err
 	}
+
+	extended, err := repairDataSquare(parentCtx, quadrant, dah, dag, dataSquare, codec, true)
+	if err == nil || err != rsmt2d.ErrUnrepairableDataSquare {
+		return extended, err
+	}
+
+	// Trying one more time to repair dataSquare, using dah.Columns
+	quadrant, err = pickQuadrant(q, dah.ColumnRoots)
+	if err != nil {
+		return nil, err
+	}
+
+	return repairDataSquare(parentCtx, quadrant, dah, dag, dataSquare, codec, false)
+}
+
+type quadrant struct {
```

Comment:
I think idiomatic Go says this should be upper-case? Not sure.
```suggestion
type Quadrant struct {
```

Reply:
But in this case current struct will be exported. As it is related only for this package I'm not sure that making in Upper case is a good idea.